File: server/utils/validators.js
/**
 * Validators
 * Utility functions for data validation
 */

const { body, validationResult } = require('express-validator');

/**
 * Validate email format
 * @param {string} email - Email to validate
 * @returns {boolean} - Whether email is valid
 */
const isValidEmail = (email) => {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
};

/**
 * Validate password strength
 * @param {string} password - Password to validate
 * @returns {Object} - Validation result with errors if any
 */
const validatePasswordStrength = (password) => {
  const errors = [];
  
  if (!password || password.length < 8) {
    errors.push('Password must be at least 8 characters long');
  }
  
  if (!/[A-Z]/.test(password)) {
    errors.push('Password must contain at least one uppercase letter');
  }
  
  if (!/[a-z]/.test(password)) {
    errors.push('Password must contain at least one lowercase letter');
  }
  
  if (!/[0-9]/.test(password)) {
    errors.push('Password must contain at least one number');
  }
  
  return {
    isValid: errors.length === 0,
    errors
  };
};

/**
 * Validate hex color code
 * @param {string} color - Color code to validate
 * @returns {boolean} - Whether color is valid
 */
const isValidHexColor = (color) => {
  return /^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/.test(color);
};

/**
 * Validate file extension against allowed types
 * @param {string} filename - Filename to validate
 * @param {string[]} allowedExtensions - Array of allowed extensions
 * @returns {boolean} - Whether file extension is valid
 */
const isValidFileExtension = (filename, allowedExtensions) => {
  if (!filename || !allowedExtensions || !Array.isArray(allowedExtensions)) {
    return false;
  }
  
  const extension = filename.split('.').pop().toLowerCase();
  return allowedExtensions.includes(extension);
};

/**
 * Common validation chains for express-validator
 */
const validationChains = {
  // User validation
  email: body('email')
    .isEmail()
    .withMessage('Please provide a valid email address')
    .normalizeEmail(),
  
  password: body('password')
    .isLength({ min: 8 })
    .withMessage('Password must be at least 8 characters long')
    .matches(/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)/)
    .withMessage('Password must contain at least one uppercase letter, one lowercase letter, and one number'),
  
  displayName: body('displayName')
    .isLength({ min: 2, max: 50 })
    .withMessage('Display name must be between 2 and 50 characters'),
  
  // Module validation
  moduleName: body('name')
    .notEmpty()
    .withMessage('Module name is required')
    .isLength({ min: 1, max: 100 })
    .withMessage('Module name must be between 1 and 100 characters'),
  
  moduleDescription: body('description')
    .optional()
    .isLength({ max: 500 })
    .withMessage('Description must be at most 500 characters'),
  
  moduleColor: body('color')
    .optional()
    .matches(/^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/)
    .withMessage('Color must be a valid hex color code'),
  
  // Note validation
  noteTitle: body('title')
    .optional()
    .isLength({ max: 200 })
    .withMessage('Title must be at most 200 characters'),
  
  // ID validation
  moduleId: body('moduleId')
    .notEmpty()
    .withMessage('Module ID is required'),
  
  noteId: body('noteId')
    .notEmpty()
    .withMessage('Note ID is required')
};

/**
 * Validate request data and return errors if any
 * @param {Object} req - Express request object
 * @returns {Object|null} - Validation errors or null if valid
 */
const validateRequest = (req) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return {
      success: false,
      errors: errors.array()
    };
  }
  return null;
};

module.exports = {
  isValidEmail,
  validatePasswordStrength,
  isValidHexColor,
  isValidFileExtension,
  validationChains,
  validateRequest
};

File: frontend/lib/utils/validators.dart
class Validators {
  /// Validates an email address
  static String? validateEmail(String? value) {
    if (value == null || value.isEmpty) {
      return 'Email is required';
    }
    
    // Simple email regex pattern
    final emailRegex = RegExp(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$');
    
    if (!emailRegex.hasMatch(value)) {
      return 'Enter a valid email address';
    }
    
    return null;
  }
  
  /// Validates a password
  static String? validatePassword(String? value) {
    if (value == null || value.isEmpty) {
      return 'Password is required';
    }
    
    if (value.length < 8) {
      return 'Password must be at least 8 characters long';
    }
    
    // Check for at least one uppercase letter
    if (!RegExp(r'[A-Z]').hasMatch(value)) {
      return 'Password must contain at least one uppercase letter';
    }
    
    // Check for at least one number
    if (!RegExp(r'[0-9]').hasMatch(value)) {
      return 'Password must contain at least one number';
    }
    
    return null;
  }
  
  /// Validates the confirm password field
  static String? validateConfirmPassword(String? value, String password) {
    if (value == null || value.isEmpty) {
      return 'Please confirm your password';
    }
    
    if (value != password) {
      return 'Passwords do not match';
    }
    
    return null;
  }
  
  /// Validates a user's display name
  static String? validateName(String? value) {
    if (value == null || value.isEmpty) {
      return 'Name is required';
    }
    
    if (value.length < 2) {
      return 'Name must be at least 2 characters long';
    }
    
    if (value.length > 50) {
      return 'Name must be less than 50 characters long';
    }
    
    return null;
  }
  
  /// Validates a module title
  static String? validateModuleTitle(String? value) {
    if (value == null || value.isEmpty) {
      return 'Title is required';
    }
    
    if (value.length > 100) {
      return 'Title must be less than 100 characters long';
    }
    
    return null;
  }
  
  /// Validates a note title
  static String? validateNoteTitle(String? value) {
    if (value == null || value.isEmpty) {
      return 'Title is required';
    }
    
    if (value.length > 200) {
      return 'Title must be less than 200 characters long';
    }
    
    return null;
  }
  
  /// Validates a tag name
  static String? validateTagName(String? value) {
    if (value == null || value.isEmpty) {
      return 'Tag name is required';
    }
    
    if (value.length > 50) {
      return 'Tag name must be less than 50 characters long';
    }
    
    return null;
  }
  
  /// Validates a color hex code
  static String? validateHexColor(String? value) {
    if (value == null || value.isEmpty) {
      return null; // Color is optional
    }
    
    // Check if value is a valid hex color
    final hexColorRegex = RegExp(r'^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$');
    
    if (!hexColorRegex.hasMatch(value)) {
      return 'Invalid color format. Use #RRGGBB or #RGB';
    }
    
    return null;
  }
  
  /// Validates a search query
  static String? validateSearchQuery(String? value) {
    if (value == null || value.isEmpty) {
      return 'Search query is required';
    }
    
    if (value.length < 2) {
      return 'Search query must be at least 2 characters long';
    }
    
    return null;
  }
}